kernel/src/engine/arrow_utils.rs

```diff
@@ -1,89 +1,895 @@
-        // we assume that in get_requested_indicies we will have caught any column name mismatches,
-        // so here we can just say that if we request the same # of columns as the parquet file
-        // actually has, we don't need to mask anything out
-        None
-    } else {
-        Some(ProjectionMask::roots(
-            parquet_physical_schema,
-            indicies.to_owned(),
-        ))
+    // TODO: Determine if it's worth checking if we're selecting everything and returning None in
+    // that case
+    Some(ProjectionMask::leaves(
+        parquet_physical_schema,
+        indices.to_owned(),
+    ))
+}
+
+/// Check if an ordering is already ordered
+fn is_ordered(requested_ordering: &[ReorderIndex]) -> bool {
+    if requested_ordering.is_empty() {
+        return true;
+    }
+    // we have >=1 element. check that the first element is ordered
+    if !requested_ordering[0].is_ordered() {
+        return false;
     }
+    // now check that all elements are ordered wrt. each other, and are internally ordered
+    requested_ordering
+        .windows(2)
+        .all(|ri| (ri[0].index() < ri[1].index()) && ri[1].is_ordered())
 }
 
+// we use this as a placeholder for an array and its associated field. We can fill in a Vec of None
+// of this type and then set elements of the Vec to Some(FieldArrayOpt) for each column
+type FieldArrayOpt = Option<(Arc<ArrowField>, Arc<dyn ArrowArray>)>;
+
 /// Reorder a RecordBatch to match `requested_ordering`. For each non-zero value in
 /// `requested_ordering`, the column at that index will be added in order to returned batch
-pub(crate) fn reorder_record_batch(
-    input_data: RecordBatch,
-    requested_ordering: &[usize],
-) -> DeltaResult<RecordBatch> {
-    if requested_ordering.windows(2).all(|is| is[0] < is[1]) {
-        // indicies is already sorted, meaning we requested in the order that the columns were
+pub(crate) fn reorder_struct_array(
+    input_data: StructArray,
+    requested_ordering: &[ReorderIndex],
+) -> DeltaResult<StructArray> {
+    if is_ordered(requested_ordering) {
+        // indices is already sorted, meaning we requested in the order that the columns were
         // stored in the parquet
         Ok(input_data)
     } else {
         // requested an order different from the parquet, reorder
-        let input_schema = input_data.schema();
-        let mut fields = Vec::with_capacity(requested_ordering.len());
-        let reordered_columns = requested_ordering
-            .iter()
-            .map(|index| {
-                fields.push(input_schema.field(*index).clone());
-                input_data.column(*index).clone() // cheap Arc clone
-            })
+        debug!("Have requested reorder {requested_ordering:#?} on {input_data:?}");
+        let num_rows = input_data.len();
+        let (input_fields, mut input_cols, null_buffer) = input_data.into_parts();
+        let mut final_fields_cols: Vec<FieldArrayOpt> = std::iter::repeat_with(|| None)
```

Comment:
Is `vec![None; requested_ordering.len()];` not work?

Reply:
Does indeed work. I for some reason didn't do it that way 🤷🏽